Build: New WRTD device and ACQ400_TR device

santorofer opened the pull request:
I have here, in this pull request, and adding: a WRTD device, and some changes to the ACQ400_TR_BASE.

The WRTD device is used to set up the WR messages used to trigger capture, including the selection of sources and signals used for that when a particular message is used as a trigger.

The ACQ400_TR_BASE class now has several changes, including:
1- ability to select between using external triggers or software trigger, or WR messages.
2- the data is recorded as an MDSplus signal, i.e data with a window and an axis.
3- calibrated signal data is included in the subnode as CAL_INPUT of the INPUT_XXX node

santorofer committed:
New WRTD device plus fixes to D-Taqc st and tr devices

santorofer committed:
Added the version of acq400_base that has now threads for some of the methods, like arm and store. Added some new nodes to TR. Making use of the new site 11 knobs in WRTD device

santorofer committed:
Remove an extra Arm() method

santorofer committed:
Added has_wr=True to use site cC

santorofer committed:
Some improvements

santorofer committed:
Using names from the uut.cC matches in the TRIG function instead of nodes values

santorofer committed:
Using new way of calling wrtd_tx_immediate. New INIT for acq2106_tr

santorofer committed:
New way to send WR immediate messages

santorofer committed:
Getting T0 trigger time from WRTD device into ACQ2106_TR device

santorofer committed:
Remove some reduntant nodes of the TR tree

santorofer committed:
Modified acq2106_tr.py to configure the signals and sources. Now _ACAQ400_TR_BASE is back to its original basics

santorofer committed:
We now pass just the WR message to wrtd_tx_immediate. Now the parts in _acq400_base are also added.

santorofer committed:
Added Pull method

santorofer committed:
Removed a hardcoded IP address left there by mistake

santorofer committed:
Some formating and cleaning

santorofer committed:
Improving initialization settings

santorofer committed:
Move initialization to NONE of the sources out of the IF statement

santorofer committed:
Simplified selection logic when WR is not used to trigger.

santorofer committed:
Added MDSplus signal to the CAL_INPUTs

santorofer committed:
Added statement to stop triggering if message doesn't match

santorofer committed:
Modifying the _ACQ400_TR_BASE to record a signal, together we all it's needed for a transient recording with or without WR

santorofer committed:
Missed adding the new acq400_base file to the previous pull request.

zack-vii commented:
perfect i was just ficxin the foiurmat string in tha one. can you check it with the changes i mad in the other pr #2065 . you need to use % style as the {} is not supported by rhel6's python.

zack-vii commented:
how does this pr relate to #1997 ? is it not affecting the same devices

zack-vii commented on pydevices/acq400Devices/acq1001_st.py:
```diff
@@ -24,11 +24,9 @@
-    acq400_base = __import__('acq400_base', globals(), level=1)
-except:
-    acq400_base = __import__('acq400_base', globals())
+import importlib
```
is importlib available for all pythons versions? then we could use it everywhere as a replacement for _minport

zack-vii commented on pydevices/acq400Devices/acq1001_tr.py:
```diff
@@ -24,11 +24,9 @@
-    acq400_base = __import__('acq400_base', globals(), level=1)
-except:
-    acq400_base = __import__('acq400_base', globals())
+import importlib
```
dito

zack-vii commented on pydevices/acq400Devices/acq2106_st.py:
```diff
@@ -24,11 +24,9 @@
-    acq400_base = __import__('acq400_base', globals(), level=1)
-except:
-    acq400_base = __import__('acq400_base', globals())
+import importlib
```
dito
if not you should revert this

zack-vii commented on pydevices/acq400Devices/acq1001_st.py:
```diff
@@ -40,6 +38,7 @@ class _ACQ1001_ST(acq400_base._ACQ400_ST_BASE):
     list(_ACQ1001_ST.base_parts) + list(_ACQ1001_ST.st_base_parts),
     acq400_base.ACQ1001_CHANNEL_CHOICES
 )
+
```
dito

zack-vii commented on pydevices/acq400Devices/acq2106_tr.py:
```diff
@@ -23,11 +23,11 @@
 
+import MDSplus
+import threading
+import importlib
```
same here

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -23,52 +23,49 @@
-import inspect
-
-import numpy as np
+from queue import Queue, Empty
```
```
if sys.version_info < (3,):
    from Queue import Queue, Empty
else:
    from queue import Queue, Empty
```
better import the package:
```
if sys.version_info < (3,):
    import Queue as queue
else:
    import queue
```

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -92,7 +89,32 @@ def init(self):
-        uut.s0.transient = "POST={}".format(self.samples.data())
+        # If PRE samples different from zero
+        print("PRE={} POST={}".format(self.presamples.data(), self.samples.data()))
+        uut.s0.transient = "PRE={} POST={}".format(self.presamples.data(), self.samples.data())
```
"{}".format(val) style will break on rhel6, please use % style

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -23,52 +23,49 @@
-    from queue import Queue, Empty
-else:
-    from Queue import Queue, Empty
+try:
```
Dont import acq400_hapi here. import it as regular package
```
import acq400_hapi
```
 in the methods you use it.. acq400_hapi is a third party package just like pexpect, serial, etc. hence it should be placed in PYTHONPATH

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -92,7 +89,32 @@ def init(self):
+        clockdiv      = mb_freq/self.freq.data()
+
+        #The following will set the sample rate, by setting the clock div.
+        uut.s1.CLKDIV = "{}".format(clockdiv)
```
more {} format

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -142,7 +164,21 @@ def __init__(self,dev):
+                if ch.on:
+                    ch.EOFF.putData(float(eoff[ic]))
+                    ch.ESLO.putData(float(eslo[ic]))
+                    expr = "{} * {} + {}".format(ch, ch.ESLO, ch.EOFF)
```
since you want to compile an mdsplus expression you should not use this. use:
```
ch.tree.tdiCompile("$ * $ + $", ch, ch.ESLO, ch.EOFF)
```

zack-vii replied:
Add you forgot to put the actual data in ch.. beast remove the EOFF ESLO and CAL_stuff and instead make teh data a signal with 
```
for ic, ch in enumerate(self.chans):
    if ch.on:
        cal = MDSplus.TdiCompile("$VALUE * $ + $", float(eslo[ic]), float(eoff[ic]))
        ch.record = MDSplus.Signal(cal, chan_data[ci])
```

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -284,18 +320,132 @@ class _ACQ400_TR_BASE(_ACQ400_BASE):
+    def stop(self):
+        import acq400_hapi
+        uut = acq400_hapi.Acq400(self.node.data())
+        print("PRE {}, POST {} and ELAPSED {}".format(uut.pre_samples(), uut.post_samples(), uut.elapsed_samples()))
```
more more

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -284,18 +320,132 @@ class _ACQ400_TR_BASE(_ACQ400_BASE):
+        thread = threading.Thread(target = self._store)
+        thread.start()
+        return None
+    STORE=store
```
no need for this.. specify the lowercase name in the METHOD of the action

in general

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -306,17 +456,172 @@ def store(self):
+                signal = MDSplus.Signal(MDSplus.Data.compile(expr), None, dim)
+                ch.CAL_INPUT.putData(signal)
+
+    _STORE=_store
```
definetly not required the uppercasing was only for old ACTION methods

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2017, Massachusetts Institute of Technology All rights reserved.
+# Copyright (c) 2020, Massachusetts Institute of Technology All rights reserved.
```
rename the module with a leading _ this will skip the whole file when parsing for Devices -> better performance (#2065 )

zack-vii commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -254,26 +289,27 @@ def run(self):
                     except socket.error as e:
                         # Something else happened, handle error, exit, etc.
-                        print("socket error", e)
+                        self.dprint(0, "socket error: %s", e)
```
👍

santorofer committed:
Remove stl file usage. Added ST highway signals settings

santorofer committed:
Added a thread for the trig function in WRTD. Separated the clock configuration from the main 400_base to the specific files for each device.

santorofer committed:
Removed commented out section and hardcoded path
Removed hardcoded path to a STL table and replace it with a tree node reference

petermilne commented on pydevices/HtsDevices/acq2106_WRPG.py:
```diff
@@ -0,0 +1,188 @@
+import os
+from tempfile import mkstemp
+
+class ACQ2106_WRPG(MDSplus.Device):
```
This is a leaf class, hard coded for the DIO482 in site 6.
We think it would be better to have an intermediate class, then up to 6 leaf classes that are almost empty, with the DIO site hardcoded at 1,2,3,4,5,6 etc. Of course, chances it always _is_ a solo module in site 6, so then it's fine how it is.

petermilne commented on pydevices/HtsDevices/acq2106_WRTD.py:
```diff
@@ -0,0 +1,157 @@
+        time.sleep(1)
+
+        wrtdtx = message
+        uut.s0.wrtd_tx_immediate = wrtdtx
```
This would be sufficient to replace 145..149..:
uut.s0.wrtd_tx_immediate = message

petermilne commented on pydevices/acq400Devices/acq1001_st.py:
```diff
@@ -24,22 +24,28 @@
+
+        uut.s0.set_knob('SYS_CLK_FPMUX', 'ZCLK')
+        freq = int(self.freq.data())
+        uut.s1.set_knob('ACQ43X_SAMPLE_RATE', "%d"%freq)
```
This is specific to ACQ435, won't work for ACQ423 ..
Please use sync_role to set clock and trigger, it's agnostic.
Also, ZCLK wouldn't be the best choice for ACQ435, XCLK is better (gives round number Hz)

petermilne replied:
Worse, acq400_base already set the clocks using sync_role. This could be harmless at best, fighting against it at worst

petermilne commented on pydevices/acq400Devices/acq1001_tr.py:
```diff
@@ -24,22 +24,29 @@
     """
     D-Tacq ACQ1001 transient support.
     """
+    def init(self):
```
again, please don't do this. It's already handled in _ACQ400_BASE.init using sync_role()

petermilne commented on pydevices/acq400Devices/acq2106_st.py:
```diff
@@ -23,24 +23,48 @@
     """
     D-Tacq ACQ2106 stream support.
     """
+    def init(self):
```
This is all very WR specific. Not all ACQ2106_ST are WR devices. We think it should be abstracted to a WR specific class with WR specific nodes. Also, we should consider what is the defaults with a WR devices, because likely most of this just isn't needed because it's set every time in a WR.

petermilne commented on pydevices/acq400Devices/acq2106_st.py:
```diff
@@ -23,24 +23,48 @@
+        clockdiv      = mb_freq/self.freq.data()
 
+        #The following will set the sample rate, by setting the clock div.
+        uut.s1.CLKDIV = "{}".format(clockdiv)
```
CLKDIV is relevant for ACQ42x, ACQ43x (maybe) but it's certainly NOT relevant for ACQ48x.. just saying..

petermilne commented on pydevices/acq400Devices/acq2106_st.py:
```diff
@@ -23,24 +23,48 @@
+        self.dprint(1, "After setting the sample rate the value in the ACQ is%r Hz", acq_sample_freq)
+
+        #The following is what the ACQ script called "/mnt/local/set_clk_WR" does to set the WR clock to 20MHz
+        uut.s0.SYS_CLK_FPMUX     = 'ZCLK'
```
So we said that sync_role was the universal clock/trig sync controller. Yes, for systems with HDMI clk/trg highway.
That actually doesn't apply for WR (blows HDMI away).
So I think we should have a parameter IS_WR and, if set, some ACQ2106 specific init does clock things. 
Also, the clock in WR is going to be very specific. We're likely to propose:
- all WR/TIGA : sets a 10MHz Sample clock (believe it or not the FPGA sets this as a constant). The ACQ423 uses CLKDIV to get its clock.
- all other WR will set a 40MHz Sample clock
All initialisation is built into the boot up, there's nothing for client software to touch.

petermilne replied:
Here's the plan. We're going to add a sync_role command to the WR systems. Now the SAME command DOES work everywhere.
On WR,  there's no concept of master, slave, fpmaster, rpmaster, so they all do the same thing, aka set your clock rate which is what you want.

petermilne commented on pydevices/acq400Devices/acq2106_tr.py:
```diff
@@ -23,26 +23,52 @@
+
+        #Set the Sampling rate in the ACQ:
+        # MB Clock (WR Clock): 20MHz
+        # mb_freq = uut.s0.SIG_CLK_MB_FREQ
```
AS previous, IS_WR ... if it "IS_WR" then CLK.d1 is 40MHz unless TIGA when it's 10MHz
(they wouldn't let me have 40MHz...)

petermilne reviewed (commented):
Hi Fernando, we're just learning how to do reviews. Our comments here.

Take away: we need to finish of the WR clock setup so that it's consistent with the other products.
ie sync_role ROLE Fs  works everywhere, (all carriers, all modules, then your code can be a LOT simpler)

petermilne commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -137,12 +188,27 @@ def __init__(self,dev):
-            self.segment_bytes = self.seg_length*self.nchans*np.int16(0).nbytes
+            self.segment_bytes = self.seg_length*self.nchans*numpy.int16(0).nbytes
+
+            #Fechting all calibration information from every channel. Save it in INPUT_XXX:CAL_INPUT
```
Fetching

petermilne commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -284,18 +350,134 @@ class _ACQ400_TR_BASE(_ACQ400_BASE):
+            # TRIGGER setting
+            if str(self.trig_src.data()) in srcs_1:
+                # uut.s0.SIG_SRC_TRG_1   = 'STRIG'
+                uut.s0.SIG_SRC_TRG_1   = str(self.trig_src.data())
+                #Setting the signal (dX) to use for ACQ2106 stream control
+                uut.s1.TRG       = 'enable'
+                uut.s1.TRG_DX    = 'd1'
+                uut.s1.TRG_SENSE = 'rising'
+
+                #EVENT0 setting in d0:
+                uut.s0.SIG_SRC_TRG_0   = str(self.event0_src.data())
+                uut.s1.EVENT0       = 'enable'
+                uut.s1.EVENT0_DX    = 'd0'
+                uut.s1.EVENT0_SENSE = 'rising'
+                uut.s0.SIG_EVENT_SRC_0 = 'TRG' # In the EVENT bus, the source needs to be TRG to make the transition PRE->POST
+
+            elif str(self.trig_src.data()) in srcs_0:
+                # uut.s0.SIG_SRC_TRG_0   = 'EXT'
+                uut.s0.SIG_SRC_TRG_0   = str(self.trig_src.data())
+                #Setting the signal (dX) to use for ACQ2106 stream control
+                uut.s1.TRG       = 'enable'
+                uut.s1.TRG_DX    = 'd0'
+                uut.s1.TRG_SENSE = 'rising'
+
+                #EVENT0 setting in d1:
+                uut.s0.SIG_SRC_TRG_0   = str(self.event0_src.data())
+                uut.s1.EVENT0       = 'enable'
+                uut.s1.EVENT0_DX    = 'd1'
+                uut.s1.EVENT0_SENSE = 'rising'
+                uut.s0.SIG_EVENT_SRC_0 = 'TRG' # In the EVENT bus, the source needs to be TRG to make the transition PRE->POST
+
+        #If WR is used to trigger, then:
+        else:
+            uut.s1.TRG       = 'enable'
+            uut.s1.TRG_DX    = str(self.trig_dx.data())
+            uut.s1.TRG_SENSE = 'rising'
+
+            uut.s1.EVENT0       = 'enable'
+            uut.s1.EVENT0_DX    = str(self.event0_dx.data())
+            uut.s1.EVENT0_SENSE = 'rising'
```
again, we think we can eliminate all/most of this stuff with a WR-capable sync role.

petermilne commented on pydevices/acq400Devices/acq400_base.py:
```diff
@@ -108,10 +106,62 @@ class _ACQ400_ST_BASE(_ACQ400_BASE):
+        super(_ACQ400_ST_BASE, self).init()
+
+        uut = acq400_hapi.Acq400(self.node.data(), monitor=True)
+        
+        # Initializing Sources to NONE:
+        # D0 signal:
+        uut.s0.SIG_SRC_TRG_0   = 'NONE'
+        # D1 signal:
+        uut.s0.SIG_SRC_TRG_1   = 'NONE'
+
+        #Trigger sources choices:
+        # d0:
+        srcs_0 = ['EXT', 'HDMI', 'HOSTB', 'GPG0', 'DSP0', 'WRTT0', 'nc', 'NONE']
+        # d1:
+        srcs_1 = ['STRIG', 'HOSTA', 'HDMI_GPIO', 'GPG1', 'DSP1', 'FP_SYNC', 'WRTT1', 'NONE']
+
+        # When WR is not being used to trigger:
+        if self.wr.data() == 0:
+            # No PRE samples --> EVENT0 doesn't need to be set.
+            # TRIGGER setting
+            if str(self.trig_src.data()) in srcs_1:
+                uut.s0.SIG_SRC_TRG_1   = str(self.trig_src.data())
+                #Setting the signal (dX) to use for ACQ2106 stream control
+                uut.s1.TRG       = 'enable'
+                uut.s1.TRG_DX    = 'd1'
+                uut.s1.TRG_SENSE = 'rising'
+
+            elif str(self.trig_src.data()) in srcs_0:
+                uut.s0.SIG_SRC_TRG_0   = str(self.trig_src.data())
+                #Setting the signal (dX) to use for ACQ2106 stream control
+                uut.s1.TRG       = 'enable'
+                uut.s1.TRG_DX    = 'd0'
+                uut.s1.TRG_SENSE = 'rising'
+            #pass
+        
+        #If WR is used to trigger, then:
+        else:
+            uut.s1.TRG       = 'enable'
+            uut.s1.TRG_DX    = str(self.trig_dx.data())
+            uut.s1.TRG_SENSE = 'rising'
```
#127 .. #158 sync_role is meant to hide ALL this stuff.
In the #142 if self.wr.data() == 0: section, sync_role does this NOW

For the #161 else: section,
We owe you a WR-specific sync_role
Also, it's going to be VERY simple, hard codes
uut.s0.SIG_SRC_TRG_1 = WRTT0
uut.s0.SIG_SRC_TRG_1 = WRTT1

=> the higher level UUT-side command saves 27 lines of confusing stuff here..

petermilne reviewed (commented):
Looks good, but there's a lot of setting clocks "by hand" in here that we don't think is necessary.
We'll implement sync_role for WR in the next week and present that.

santorofer committed:
Added but the call to setevent() that was removed for testing inside the box

santorofer committed:
New WR dedicated subclasses for ST and TR. Simplify trigger logic for ST. Be sure CLK get set once for either WR or not WR devices

seanalsop commented on pydevices/acq400Devices/acq1001_st.py:
```diff
@@ -24,22 +24,27 @@
     """
     D-Tacq ACQ2106 stream support.
     """
+    def init(self):
+        super(_ACQ1001_ST, self).init()
+        uut = acq400_hapi.Acq400(self.node.data(), monitor=False)
+
+        uut.s0.set_knob('SYS_CLK_FPMUX', 'XCLK')
```
The parent class is ACQ400_ST_BASE, which inherits 'init' from ACQ400_BASE. Creating a new init here would replace that init function. This init doesn't do much anyway, better removed I think.

seanalsop commented:
Peter and I are working on a WR unified sync_role. We will suggest changes to init functions once this has been done.

santorofer commented:
Hi Sean, thank you. Yes, excellent.

santorofer committed:
Preparing for the WR sync_role

santorofer committed:
Improving transient init function. This will change with new WR sync_role.

santorofer committed:
Better handling of the site number for a DIO module

santorofer commented:
Dear Sean and Peter,

I was wondering if you had the chance to work on the  WR-aware sync_role?

Cheers!
Fernando

> On Aug 24, 2020, at 9:59 AM, seanalsop <notifications@github.com> wrote:
> 
> 
> Peter and I are working on a WR unified sync_role. We will suggest changes to init functions once this has been done.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/MDSplus/mdsplus/pull/2066#issuecomment-679178553>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKW7QGINJHDZOWYOGFAIOILSCJ54TANCNFSM4OBGIIMA>.
>

petermilne commented:
Hello Fernando

Good point. So I did code it up 23 days ago:
https://github.com/petermilne/acq400ioc/blob/epics7/CARE/SITE0/sync_role.wr

It's there in the current release:
acq2106_262> head /mnt/VERSION
RELEASE acq400-266-20200930164230

acq2106_262> ls /usr/local/CARE/SITE0/
fp_trg_out.common     sync_role.common
mb_clk_freq_check.py  sync_role.wr

# the knob gets selected on boot

acq2106_262> ls -l /etc/acq400/0/sync_role
lrwxrwxrwx    1 root     root            47 Oct  7 15:01 /etc/acq400/0/sync_role -> /usr/local/epics/scripts/SITE0/sync_role.common

# This is a regular system, a WR system should make the link to sync_role.wr

Sorry, I haven't had time to _look_ at WR in the last month, but it's now going to get REALLY important for the next month, so, if the .wr flavour of sync_role isn't working, we'll be able to fix it.

Meanwhile, do try it on your systems. It's really a big simplification of the regular sync_role (no masters or slaves).

Cheers

Peter


On 08/10/2020 07:30, Fernando Santoro wrote:
> Dear Sean and Peter,
> 
> I was wondering if you had the chance to work on the  WR-aware sync_role?
> 
> Cheers!
> Fernando
> 
>> On Aug 24, 2020, at 9:59 AM, seanalsop <notifications@github.com> wrote:
>>
>>
>> Peter and I are working on a WR unified sync_role. We will suggest changes to init functions once this has been done.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub <https://github.com/MDSplus/mdsplus/pull/2066#issuecomment-679178553>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKW7QGINJHDZOWYOGFAIOILSCJ54TANCNFSM4OBGIIMA>.
>>
> 
> 
> 


-- 
Peter Milne
Director of Sales
www.d-tacq.com

santorofer committed:
Added sync_role for WR devices

santorofer committed:
There is no need of using is_wr to select the sync_role() to use.

santorofer committed:
Add settings for TX

santorofer committed:
Add WR trigger sources FPTRG or HDMI

santorofer committed:
Fix wrong data and add print statemets

santorofer committed:
Changed SIG_EVENT_SRC_0 to TRG

santorofer committed:
Change io_buffer_size to seg_length

santorofer committed:
Added print statement about making new buffers

santorofer committed:
Add 24bit _ACQ400_ST_BASE_24bit class for the 435ELF

santorofer committed:
Add new code to make usage of 24bits ACQ 435s

santorofer commented:
Closing, since most of the changes done in this PR has now been developed in new independent PRs. Another words, this PR has been superseded by new PRs addressing many of the changes shown here.
